File: Skywalker/MassObject.m
//
//  MassObject.m
//  Skywalker
//
//  Created by Zion on 8/8/13.
//  Copyright (c) 2013 Zion. All rights reserved.
//

#import "MassObject.h"

@implementation MassObject

-(id)copyWithZone:(NSZone *)zone
{
    MassObject *copyer=[[MassObject alloc] init];
    if (copyer)
    {
        copyer.mass=_mass;
        copyer.radius=_radius;
        copyer.velocityVector=_velocityVector;
        copyer.forceVector=_forceVector;
    }
    return copyer;
}

-(Vector)velocityAfterTimeInterval:(NSTimeInterval)timeInterval
{
    return setLength(_forceVector, timeInterval*vectorLength(_forceVector)/_mass);
}

-(void)updateVelocityAfterTimeInterval:(NSTimeInterval)timeInterval
{
    _velocityVector=[self velocityAfterTimeInterval:timeInterval];
}

-(MassObject*)collideWithMassObject:(MassObject*)colObj
{
    return [MassObject massObjectCollidedByMassObject:self and:colObj];
}

+(MassObject*)massObjectCollidedByMassObject:(MassObject*)colObj0 and:(MassObject*)colObj1
{
    
    //质量相加，体积以最大为准
    MassObject *obj_new=[[MassObject alloc] init];
    obj_new.mass=colObj0.mass+colObj1.mass;
    obj_new.radius=MAX(colObj0.radius,colObj1.radius);
    obj_new.forceVector=colObj0.forceVector;//都一样
    obj_new.velocityVector=multiplyFactor(addVector(multiplyFactor(colObj0.velocityVector, colObj0.mass),multiplyFactor(colObj1.velocityVector, colObj1.mass)),1/obj_new.mass);
    
    return obj_new;
    
}

@end

File: Skywalker/MassObject.cpp
//
//  MassObject.cpp
//  Skywalker
//
//  Created by Zion on 8/8/13.
//  Copyright (c) 2013 Zion. All rights reserved.
//

#include "MassObject.h"
#include <algorithm>

MassObject::~MassObject()
{

}

MassObject::MassObject()
:radius(0),
mass(0),
gravitationRadius(0),
gravity(0),
velocityVector(Vector::Zero),
forceVector(Vector::Zero)
{
    
}

MassObject::MassObject(const MassObject &obj)
:radius(obj.radius),
mass(obj.mass),
gravitationRadius(obj.gravitationRadius),
gravity(obj.gravity),
velocityVector(obj.velocityVector),
forceVector(obj.forceVector)
{
    
}

void MassObject::setMass(const ScientificNumber &m)
{
    mass=m;
    calculateGravity();
}

void MassObject::setRadius(const ScientificNumber &r)
{
    radius=r;
    calculateGravitationRadius();
    calculateGravity();
}

const ScientificNumber MassObject::getMass() const
{
    return mass;
}

const ScientificNumber MassObject::getRadius() const
{
    return radius;
}

const ScientificNumber MassObject::getGravitationRadius()const
{
    return gravitationRadius;
}

const ScientificNumber MassObject::getGravity()const
{
    return gravity;
}

void MassObject::calculateGravitationRadius()
{
    gravitationRadius=(radius)*4;
}

void MassObject::calculateGravity()
{
    gravity=10000;
}

void MassObject::updateVelocityAfterTimeInterval(const double timeInterval)
{
    velocityVector+=((forceVector)/(mass))*timeInterval;
}

void MassObject::collideWithMassObject(const MassObject &obj)
{
    if (radius<obj.radius)
    {
        radius=obj.radius;
    }
    velocityVector=((velocityVector)*(mass)+(obj.velocityVector)*(obj.mass))/((mass)+(obj.mass));
    mass+=(obj.mass);
}
